### PR TITLE
Sequential v2

### PR DESCRIPTION
**Pull request by rhensen**
This PR reaches the first version of many of the v1 soil and crop objectives: 

Management practices:

- Crop types - corn (grain and silage), alfalfa, grass (tall fescue_, soybeans, winter wheat, winter rye, triticale
- Cover cropping
- No-till vs. tillage vs. conservation tillage
- Manure and fertilizer application - surface, incorporation, injections

Outcomes: 

- N2O, NH3, and CO2 Emissions
- N & P Leaching and Runoff
- Water use
- Soil C
- Crop yields and compositions

In addition, unit tests are provided for some of the recent changes. We hope to add more of these in the future.

### CODE REVIEW

File: RUFAS/routines/field/field_management/field_management.py
Comment (rhensen) [inline RUFAS/routines/field/field_management/field_management.py]: percent_nutrient_incorporated
percent_nutrient_mixed

File: RUFAS/output_handler/reports/field_report.py
Comment (PooyaHekmati) [inline RUFAS/output_handler/reports/field_report.py:329]: There are two levels of repetition here, both can be mitigated using zip() and lambda functions with some loops. Also, to add clarity, this big dict structure initialization should be wrapped in a function with an adequately intuitive name.

File: RUFAS/output_handler/reports/field_report.py
Comment (PooyaHekmati) [inline RUFAS/output_handler/reports/field_report.py]: misplaced curly brace

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: Line too long, please see https://peps.python.org/pep-0008/#maximum-line-length

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:36]: there are no if conditions next to this comment, how is this decision being made?

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: remove comment by something like this:
crop_was_killed_yesterday = crop_type.killed
if crop_was_killed_yesterday:

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:45]: can we wrap this in a function and call it set_next_crop or something like that?

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unclear and unexplained hardcoded value, what is the significance of 1? what happens if we change it to 0?

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: vague and unclear comment

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unnecessary comment

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unnecessary and probably misleading comment. if crop_type.planted is true, how do we know it has been just planted? Moreover, can a type of crop really be planted? A crop can be planted, the type is describing it. A type is not plantable, a crop is.

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unnecessary comment

**Review by PooyaHekmati — COMMENTED** [review]
I made these comments as a partial review to unblock this PR. Will get back to it and add more soon.

File: RUFAS/output_handler/reports/field_report.py
Comment (HectorMM3) [inline RUFAS/output_handler/reports/field_report.py:309]: Why is M_DRP_runoff tabbed over? Does it need to be?

File: RUFAS/routines/field/crop/crop.py
Comment (HectorMM3) [reply RUFAS/routines/field/crop/crop.py:45]: Pooya, I think we want to leave it as is for now because we can going to make version 3 more complex at that time we will attempt to put it in a function.

File: RUFAS/routines/field/crop/crop.py
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop.py:137]: Did we comment this out to avoid the error in carbon? I think we still need this variable as an output.

File: RUFAS/routines/field/crop/crop.py
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop.py:339]: #replace comment with "if the crop is grown in the fall or winter then it gets placed later in the year if it is a "warm #season" -planted and grown during spring and summer or null then put it belongs in the initial year slot."

File: RUFAS/routines/field/crop/crop.py
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop.py:358]: why is it commented out?

File: RUFAS/routines/field/crop/crop_types/spring_wheat.py
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop_types/spring_wheat.py:31]: please delete "# TODO: change back" it should not be changed back spring wheat does not fix nitrogen so the False value is correct.

File: RUFAS/routines/field/field_management/manure_application.py
Comment (HectorMM3) [inline RUFAS/routines/field/field_management/manure_application.py:152]: add in 0.05 and 0.06 into the comments so we can have them for later.

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py:149]: I see a weird red symbol at the end of line 146.

**Review by HectorMM3 — CHANGES_REQUESTED** [review]
Reid, only small comments. I would proceed, and we can follow up on the details later.

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [reply RUFAS/routines/field/crop/crop.py:45]: In this case, we should create a git issue for it and add it here as TODO

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [reply RUFAS/routines/field/crop/crop.py:36]: A reviewer's comment needs to be addressed either as changes in the code or as a conversation on why the comment was wrong or not applicable. Just marking the comment resolved is not helpful.

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:170]: where is this -1 coming from?

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:229]: what about moving these comments to docstrings which ended two lines above?

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:233]: readability can be enhanced by moving len(crop.grow_regimen) == time.year or to next line, so each line checks one condition

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:293]: Please add a verb such as loading, setting, etc.

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:298]: avoid hardcoding by changing to 
double_cropping_limit_default_value = 2
double_cropping_limit = double_cropping_limit_default_value

File: RUFAS/routines/field/crop/crop.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:339]: how is this if statement being executed? where is the decision being made?

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:9]: Invalid format for docstrings

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:38]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:39]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:58]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:53]: Please add reference for these values (a link to a document which explains these)

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:87]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:104]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:122]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:143]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:162]: make git issue and mention it here

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:163]: make git issue and put it here

File: RUFAS/routines/field/crop/crop_types/soybean.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/soybean.py:102]: this explanation should be inside docstrings

File: RUFAS/routines/field/crop/crop_types/triticale.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/triticale.py:9]: invalid format for docstring

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py]: what about residue_incorp = 0 ?

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py
Comment (PooyaHekmati) [reply RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py:149]: That shows that this is the end of the file (EOF) and no new lines are after this point.

File: RUFAS/routines/field/soil/soil.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/soil/soil.py:95]: why is this commented out?

File: RUFAS/routines/field/soil/soil_erosion.py
Comment (PooyaHekmati) [inline RUFAS/routines/field/soil/soil_erosion.py:51]: why is this commented out?

File: config/defintions.py
Comment (PooyaHekmati) [inline config/defintions.py:2]: why is this needed? where is it used?

**Review by PooyaHekmati — CHANGES_REQUESTED** [review]

**Comment by PooyaHekmati** [comment]
There are a bunch of TODOs in this PR, I commented on some of them but gave up on commenting on them all. A TODO is not desired, it is only acceptable when a git issue is created for it and it is mentioned in front of the TODO. Also, there are a bunch of docstrings with invalid formats. More importantly, this PR is heavily missing unit tests.

**Comment by KFosterReed** [comment]
I tried to run this and got a Key Error for the alfalfa 'extracted' input. It looks like you added this input variable to the alfalfa crop but it isn't added to all the crop jsons. Here is the full error report:

 

          > Initializing simulation environment from input\ARL.json
          > Traceback (most recent call last):
          >   File "C:\Users\kfr3\PycharmProjects\MASM\main.py", line 25, in <module>
          >     main()
          >   File "C:\Users\kfr3\PycharmProjects\MASM\main.py", line 20, in main
          >     simulator = SimulationEngine(input_file_path)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\simulation_engine.py", line 23, in __init__
          >     self._initialize_simulation(input_file_path)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\simulation_engine.py", line 153, in _initialize_simulation
          >     self.state = State(data['farm'], self.config,
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\classes.py", line 46, in __init__
          >     self.fields = Fields(data['fields'], time)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\field.py", line 45, in __init__
          >     self.fields[field_name] = Field(field_name, field_data, time)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\field.py", line 225, in __init__
          >     self.crop = Crop(crop_data, time)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\crop\crop.py", line 263, in __init__
          >     crop = alfalfa.Alfalfa(crop_name, crop_data)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\crop\crop_types\alfalfa.py", line 34, in __init__
          >     self.extracted = alfalfa_data['extracted']
          > KeyError: 'extracted'
          > 
          > Process finished with exit code 1
  
I got a similar key error for a tillage attribute of soil

File: RUFAS/routines/field/soil/soil.py
Comment (KFosterReed) [reply RUFAS/routines/field/soil/soil.py:95]: I think this is supposed to be deleted. Is that correct?

File: RUFAS/routines/field/soil/soil_erosion.py
Comment (KFosterReed) [reply RUFAS/routines/field/soil/soil_erosion.py:51]: Same as the other practice factor. Please delete all references

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop_types/cereal_rye.py:9]: I think we should make a GIT issue to reformat the documentation in all crop types. They are all following this same style of in code documentation so let's change them all at one time and not let that hold up this pull request

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop_types/cereal_rye.py:38]: Ideally they are GIT issues but let's at least have information about what the TODOs mean somewhere. As Pooya mentioned in the meeting, they can be in a google doc, written here in the comment or made into a git issue but I agree that we don't want to pull in even more cryptic TODOs than we already have

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop_types/cereal_rye.py:53]: If the reference values are provided in the pseudocode documentation then values like this are OK for now

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py
Comment (KFosterReed) [reply RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py]: this makes sense to me!

File: config/defintions.py
Comment (KFosterReed) [reply config/defintions.py:2]: Also curious...

File: RUFAS/output_handler/reports/field_report.py
Comment (KFosterReed) [reply RUFAS/output_handler/reports/field_report.py:329]: I think we should make a GitIssue for updating the output handlers in general and leave this as is for this pull request

File: RUFAS/routines/field/crop/crop.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py:45]: It will be good to translate these crop management functionalities into functions but I agree with Hector that the higher priority is to pull this into the master branch

File: RUFAS/routines/field/crop/crop.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py]: I also find the logic here and below hard to follow for most of the daily_reset() function. Providing pseudocode (and possibly a diagram?) would be super helpful

File: RUFAS/routines/field/crop/crop.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py:170]: I am also confused about the logic here. It would be helpful to have the variables used here defined in the Sphinx documentation above & in the pseudocode (or at least one or the other).

File: RUFAS/routines/field/crop/crop.py
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py:229]: There are several places like this where the code can be improved for readability. I think tagging them with a TODO and linking them into a single GITissue would be great

**Review by KFosterReed — CHANGES_REQUESTED** [review]
This seems very close to me but there are a few things that need to be done before we can merge it in: 
1) The logic of how the crop rotations are managed needs to be more clear. That is the heart of the new functionality and it is very difficult to follow. Whether it is with external documentation, updates to the in code documentation, or changes to the code itself does not matter to me. 
2) Please create git issues that group opportunities for improving the code clarity and 'cleanness' and add TODOs where identified
3) The model needs to run on all .json files

**Comment by KFosterReed** [comment]
@rhensen I am guessing the new PR replaces this one? If so, can we close this one?

**Pull request closed without merging**
